py4DSTEM/process/phase/parameter_optimize.py

```diff
@@ -0,0 +1,449 @@
+
+            # Create affine transform object
+            tr = affine(**afft_args)
+            # Apply affine transform to pixel grid, using the
+            # calibrations lifted from the dataset
+            dataset = init_static_args["datacube"]
+            init_args["initial_scan_positions"] = self._get_scan_positions(tr, dataset)
+
+            ptycho = obj(**init_args)
+            prep(ptycho, **prep_args)
+            recon(ptycho, **reco_args)
+
+            return np.log(ptycho.error)
+
+        return f
+
+    def _set_optimizer_defaults(self):
+        """
+        Set all of the verbose and plotting to False
+        """
+        self._init_static_args["verbose"] = False
+
+        self._preprocess_static_args["plot_center_of_mass"] = False
+        self._preprocess_static_args["plot_rotation"] = False
+        self._preprocess_static_args["plot_probe_overlaps"] = False
+
+        self._reconstruction_static_args["progress_bar"] = False
+        self._reconstruction_static_args["store_iterations"] = False
+        self._reconstruction_static_args["reset"] = True
+
+
+class OptimizationParameter:
+    """
+    Wrapper for scikit-optimize Space objects used for convenient calling in the PtyhochraphyOptimizer
+    """
+
+    def __init__(
+        self,
+        initial_value: Union[float, int, bool],
+        lower_bound: Union[float, int, bool] = None,
+        upper_bound: Union[float, int, bool] = None,
+        scaling: str = "uniform",
+        space: str = "real",
+        categories: list = [],
+    ):
+        """
+        Wrapper for scikit-optimize Space objects used as inputs to PtychographyOptimizer
+
+        Parameters
+        ----------
+        initial_value:
+            Initial value, used for first evaluation in optimizer
+        lower_bound, upper_bound:
+            Bounds on real or integer variables (not needed for bool or categorical)
+        scaling: str
+            Prior knowledge on sensitivity of the variable. Can be 'uniform' or 'log-uniform'
+        space: str
+            Type of variable. Can be 'real', 'integer', 'bool', or 'categorical'
+        categories: list
+            List of options for Categorical parameter
+        """
+        # Check input
+        space = space.lower()
+        assert space in (
+            "real",
+            "integer",
+            "bool",
+            "categorical",
+        ), f"Unknown Parameter type: {space}"
+
+        scaling = scaling.lower()
+        assert scaling in ("uniform", "log-uniform"), f"Unknown scaling: {scaling}"
+
+        # Get the right scikit-optimize class
+        space_map = {
+            "real": Real,
+            "integer": Integer,
+            "bool": Categorical,
+            "categorical": Categorical,
+        }
+        param = space_map[space]
+
+        # If a boolean property, the categories are True/False
+        if space == "bool":
+            categories = [True, False]
+
```

Comment:
Assert is really for debugging, not catching user ValueErrors in my book (?)

```suggestion

        if not categories:
            raise ValueError("Empty list of categories!")
```